### PR TITLE
DRAFT : [Meta Provider] draft Metadata Manager

### PR DESCRIPTION
**Pull request by gandi0330**
draft Metadata Manager code.

Related issue https://github.com/Samsung/ONE-vscode/issues/1277

### CODE REVIEW

File: src/MetadataManager/EventManager.ts [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+    metadata[relativePath] = {};
+    metadata[relativePath]["name"] = filename;
+
+    metadata[relativePath]["fileExtension"] = filename.split(".").at(-1);
+    metadata[relativePath]["createTime"] = stats.birthtime;
+    metadata[relativePath]["modifiedTime"] = stats.mtime;
+    metadata[relativePath]["isDeleted"] = false;
+
+    await Metadata.setMetadata(afterhash, metadata);
```
Comment (dayo09) [inline src/MetadataManager/EventManager.ts:176]: As of our offline conversation, you can extract this as a `Metadata`'s function or so.

File: src/MetadataManager/pathToHash.ts [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+import * as vscode from 'vscode';
+import * as crypto from 'crypto';
+import fs from 'fs';
+import {Metadata} from './metadataAPI';
+import {MetadataEventManager} from './EventManager';
+
+export class PathToHash{
+    private static instance: PathToHash;
+    private pathToHash: any;
```
Comment (dayo09) [inline src/MetadataManager/pathToHash.ts:25]: (trivial) Putting an underscore makes it easy to distinguish private member variables.

Maybe naming pathToHash as 'map' can help readers to notice that it's a internal map data.
```suggestion
    private static _instance: PathToHash;
    private _map: any;
```

File: src/MetadataManager/pathToHash.ts [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    public static async getInstance() {
+        if (!this.instance) {
+            this.instance = new PathToHash();
+            this.instance.pathToHash = await this.instance.initPathToHash();
+            await this.instance.validateMetadata(this.instance.pathToHash);
+        }
+        return this.instance;
+    }
+
+    private async initPathToHash() {
```
Comment (dayo09) [inline src/MetadataManager/pathToHash.ts:39]:
```suggestion
    private async init() {
```

There is no need to repeat the classname 'PathToHash'

File: src/MetadataManager/pathToHash.ts [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+                        metadata[key].isDeleted = true;
+                    }
+                }
+                Metadata.setMetadata(hash, metadata);
+            }
+        }
+    }
+
+
+    public getPathToHash(uri: vscode.Uri ) {
```
Comment (dayo09) [inline src/MetadataManager/pathToHash.ts:173]:
```suggestion
    public get(uri: vscode.Uri ) {
```


There is no need to repeat the classname 'PathToHash' (ditto)

File: src/MetadataManager/pathToHash.ts [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+            }
+            pathToHash = pathToHash[data];
+        });
+
+        return pathToHash;
+    }
+
+
+    private async generateHash(uri: vscode.Uri) {        
+        return  crypto.createHash('sha256').update(Buffer.from(await vscode.workspace.fs.readFile(uri)).toString()).digest('hex');
```
Comment (dayo09) [inline src/MetadataManager/pathToHash.ts:191]: Could you write down this TODO?
```
// TODO: Error handling
```

File: src/MetadataManager/pathToHash.ts [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            // already deleted
+            return;
+        }
+        delete obj[paths[paths.length-1]];
+        if (paths.length > 1) {
+            this.deleteEmptyFolder(this.pathToHash, paths, 0);
+        }
+    }
+
+    private deleteEmptyFolder(parent: any, paths: string[], idx: number) {
```
Comment (dayo09) [inline src/MetadataManager/pathToHash.ts:264]: it's not actually deleting a folder, maybe 'markDeletedFolder' is better..?

**Comment by gandi0330** [comment]
I reflected all of the previous code reviews

**Pull request closed without merging**